janus_server/src/aggregator.rs

```diff
@@ -622,4 +1018,372 @@ mod tests {
+            Arc::new(datastore),
+            clock,
+            skew,
+            Role::Helper,
+            verify_param,
+            hpke_recipient,
+            generate_hmac_key(),
+        )
+        .unwrap();
+
+        let result = warp::test::request()
+            .method("POST")
+            .path("/aggregate")
+            .body(AuthenticatedEncoder::new(request).encode(&generate_hmac_key()))
+            .filter(&filter)
+            .await;
+
+        if let Err(rejection) = result {
+            assert_matches!(
+                rejection.find::<Error>(),
+                Some(Error::MessageDecode(CodecError::Other(_)))
+            );
+        } else {
+            panic!("Should get rejection");
+        }
+    }
+
+    #[tokio::test]
+    async fn aggregate_init() {
+        // Prepare datastore & request.
+        install_trace_subscriber();
+
+        let task_id = TaskId::random();
+        let vdaf = Prio3Aes128Count::new(2).unwrap();
+        let verify_param = vdaf.setup().unwrap().1.remove(1);
+        let (datastore, _db_handle) = ephemeral_datastore().await;
+        let clock = MockClock::default();
+        let skew = Duration::minutes(10);
+        let hpke_recipient =
+            HpkeRecipient::generate(task_id, Label::InputShare, Role::Client, Role::Helper);
+        let hmac_key = generate_hmac_key();
+
+        datastore
+            .run_tx(|tx| {
+                Box::pin(async move {
+                    tx.put_task(&TaskParameters::new_dummy(task_id, Vec::new()))
+                        .await
+                })
+            })
+            .await
+            .unwrap();
+
+        // report_share_0 is a "happy path" report.
+        let input_share = generate_helper_input_share(&vdaf, &(), &0);
+        let report_share_0 = generate_helper_report_share::<Prio3Aes128Count>(
+            task_id,
+            generate_nonce(&clock),
+            &hpke_recipient.config,
+            &input_share,
+        );
+
+        // report_share_1 fails decryption.
+        let mut report_share_1 = report_share_0.clone();
+        report_share_1.nonce = generate_nonce(&clock);
+        report_share_1.encrypted_input_share.payload[0] ^= 0xFF;
+
+        // report_share_2 fails decoding.
+        let nonce = generate_nonce(&clock);
+        let mut input_share_bytes = input_share.get_encoded();
+        input_share_bytes.push(0); // can no longer be decoded.
+        let associated_data = Report::associated_data(nonce, &[]);
+        let report_share_2 = generate_helper_report_share_for_plaintext(
+            task_id,
+            nonce,
+            &hpke_recipient.config,
+            &input_share_bytes,
+            &associated_data,
+        );
+
+        // report_share_3 fails prepare_init/prepare_step, because the input share is for a
+        // different VDAF.
+        // TODO(brandon): find a better/more-blessed way to generate a bogus input share that will
+        // be detected by the VDAF in use.
```

Comment:
You're correct, of course. I created a `test_util::fake::Vdaf` class that is a minimal implementation of the `vdaf::{Vdaf,Aggregator,Client}` traits with all of the implementations required for a VDAF to be usable by `janus`, and implemented a `prepare_init` failure test (and added a `prepare_step` failure test too, since it was only a small amount of additional work).

Other than being useful for testing, this fake might be the best indicator of what implementations are practically required for a VDAF to work with `janus`. As it is, we're already at ~100 lines of code even for a do-nothing implementation!